### PR TITLE
New event: Map.ChangingIntoGrenade (my first transpiler <3)

### PR DESCRIPTION
**Pull request by galaxy119**

### CODE REVIEW

File: Exiled.Events/EventArgs/ChangingIntoGrenadeEventArgs.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+        /// <summary>
+        /// Gets or sets a value indicating what type of grenade will be spawned.
+        /// </summary>
+        public ItemType Type { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether the pickup will be changed.
+        /// </summary>
+        public bool IsAllowed { get; set; } = true;
```
Comment (iopietro) [inline Exiled.Events/EventArgs/ChangingIntoGrenadeEventArgs.cs:41]: This should be an optional parameter, like every other events with IsAllowed.

File: Exiled.Events/Handlers/Map.cs [hunk 2]
```diff
@@ -147,5 +154,11 @@ public static class Map
         /// Called after the map is generated.
         /// </summary>
         public static void OnGenerated() => Generated.InvokeSafely();
+
+        /// <summary>
+        /// Called before the server changes a <see cref="Pickup"/> into a <see cref="Grenade"/> when hit by an explosion.
+        /// </summary>
+        /// <param name="ev">The <see cref="ChangingIntoGrenadeEventArgs"/> instance.</param>
+        public static void OnServerChangingGrenade(ChangingIntoGrenadeEventArgs ev) => ChangingIntoGrenade.InvokeSafely(ev);
```
Comment (iopietro) [inline Exiled.Events/Handlers/Map.cs:162]: It should be named `OnChangingIntoGrenade`.

File: Exiled.Events/Patches/Events/Map/ChangingIntoGrenade.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            // The index offset
+            int offset = 2;
+
+            // Find the last return false call.
+            int index = newInstructions.FindLastIndex(instruction => instruction.opcode == OpCodes.Ldc_I4_0) + offset;
+
+            // Generate a continue lable.
+            Label continueLabel = generator.DefineLabel();
+
+            // Declare ServerChangingGrenadeEventArgs, to be able to store it's instance with "stloc.s".
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Map/ChangingIntoGrenade.cs:48]: You should change the comment to reflect the new name (`ChangingGrenadeEventArgs`).

File: Exiled.Events/Handlers/Map.cs (hunk 2 above)
Comment (galaxy119) [reply Exiled.Events/Handlers/Map.cs:162]: oop

File: Exiled.Events/Patches/Events/Map/ChangingIntoGrenade.cs [hunk 4]
```diff
@@ -90,8 +90,8 @@ private static IEnumerable<CodeInstruction> Transpiler(IEnumerable<CodeInstructi
                 new CodeInstruction(OpCodes.Callvirt, PropertySetter(typeof(Grenades.Grenade), nameof(Grenades.Grenade.NetworkfuseTime))),
             });
 
-            foreach (CodeInstruction instruction in newInstructions)
-                yield return instruction;
+            for (int i = 0; i < newInstructions.Count; i++)
```
Comment (iopietro) [inline Exiled.Events/Patches/Events/Map/ChangingIntoGrenade.cs:94]: I know that `i` is fine, but we used `z` for every other events.

Comment (galaxy119) [reply Exiled.Events/Patches/Events/Map/ChangingIntoGrenade.cs]: fixed

**Review by iopietro — APPROVED** [review]

**Review by iRebbok — APPROVED** [review]

**Review by iopietro — APPROVED** [review]

**Pull request merged**